scipy/sparse/csgraph/_flow.pyx

```diff
@@ -541,27 +542,60 @@ cdef ITYPE_t _augment_paths(
         An arbitrary flow in the layered graph.
 
     """
-    if current == sink:
-        return flow
-
-    cdef ITYPE_t dst_vertex, result_flow, e
-
-    while progress[current] < edge_ptr[current + 1]:
+    cdef ITYPE_t top, dst_vertex, result_flow
+    cdef ITYPE_t stack_top, e, current
+    top = 0
+    stack[top][0] = source
+    stack[top][1] = flow
+
+    while top > -1:
+        current = stack[top][0]
+        flow = stack[top][1]
         e = progress[current]
         dst_vertex = heads[e]
         if (capacities[e] > 0 and
                 levels[dst_vertex] == levels[current] + 1):
-            result_flow = _augment_paths(edge_ptr, dst_vertex, sink,
-                                         levels, heads, rev_edge_ptr,
-                                         capacities, progress, flows,
-                                         min(flow, capacities[e]))
-            if result_flow:
-                capacities[e] -= result_flow
-                capacities[rev_edge_ptr[e]] += result_flow
-                flows[e] += result_flow
-                flows[rev_edge_ptr[e]] -= result_flow
-                return result_flow
-        progress[current] += 1
+            if dst_vertex == sink:
+                result_flow = min(flow, capacities[e])
```

Comment:
Duplicated `min(flow, capacities[e])` could be avoided by calculating it outside of this `if` block.